### PR TITLE
[react-query] Update API to 1.10.0

### LINKED ISSUE #42678: Update types for react-query v1
- [x] I tried using the `@types/react-query` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript

Hi @lukaszfiszer, @jacehensley and @matteofrana, I have updated `react-query` to the latest stable version (1.0.2) but the type definitions are outdated.

### PR DESCRIPTION
**Pull request by Igorbek**
- breaking changes in API introduced in 1.0

Fixes #42678
Alternate to #42830

- [x] `useQuery`
  - [x] use variables in the query function
- [x] `usePaginatedQuery`
- [x] `useInfiniteQuery` (_type inference is not ideal; with untyped inlined functions, TS cannot infer type for 'more variable'_)
- [x] `useMutation`
- [x] `queryCache`
- [x] `useIsFetching`
- [x] `ReactQueryConfigProvider`
- [x] `setConsole`

Also added test cases from #42830. It seems a few cases that didn't work there now work here. The only missing feature is results based on discriminated unions where you can conclude `data != undefined` when `status == 'success'`. I don't think that is a good assumption and the future would worth it, because there are might be cases with the stale state.

### CODE REVIEW

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: Did you considered merging those 2 overloads into one? The difference in `TKey` arg could be handled by wrapping it with a conditional type before passing into `QueryFunction`

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: The first element of query key must always be `string` . With this type definition this code does not raise type error:

```
useQuery([10], (key) => {...} )
```

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: any reason for not using tuple type? `[string, ... QueryKeyPart[]]`?

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: remove comments

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: Although you're right that we can't conclude `data` based on `success` status, discriminated union here still make sense for `error` and `loading` statuses.

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [inline types/react-query/index.d.ts]: Can you add discrimanated untion here?

**Comment by lukaszfiszer** [comment]
Hi @Igorbek, thanks for your PR. You seems to have resolved most of problems that I have struggled with in my original PR, especially around query variables. Please take a look at my comments.

Also, for the final version I don't think it's worth keeping tests from my PR in a separate function - just include them in the main tests or remove if they seem redundant.

**Comment by lukaszfiszer** [comment]
One more thing - please check if you can use [typescript-tuple](https://www.npmjs.com/package/typescript-tuple) `Concat` instead of ts-toolbelt `List.Append` function. The former requires TS 3.0 instead of TS 3.7

File: types/react-query/index.d.ts
Comment (burtyish) [reply types/react-query/index.d.ts]: >  it can be any length of any items that are serializable
https://github.com/tannerlinsley/react-query/issues/257#issuecomment-600441958

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: I'm trying to not overuse complex type constructs such as conditional types if you can achieve the same with overloading or any other simpler constructs. TS usually works better with them.
Also, I wanted to be sure that what a key means is well defined. In this case, the key is a tuple with serializable elements.
In this overload, I probable needed to call this type `TSingleKey` or something like that.

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: I know that is written in the documentation but the implementation is really agnostic to what the key must be. It is just any serializable array. Moreover, the documentation is confusing and sometimes says contradicting things. For example, it says that the key must be `[string, object]` but have many examples where it deviates from that, such as examples where more elements or any other constituents allowed such as nested arrays. I concluded that the doc wasn't updated much while API was being changed and that requirement was from earlier versions.

Comment (Igorbek) [reply types/react-query/index.d.ts]: I actually agree that we can require the first element to be a string to make it clear that is the recommended way to go.
If you think it is better, I will make a change. Let me know.

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: This was a tradeoff. The way the key constituents are to the query function makes it hard to reconstruct the signature of the query function. This is how a query function is being called: `queryFn(...queryKey, ...queryVariables)` or in case of infinite scroll `queryFn(...queryKey, ...queryVariables, moreVariable)`. In this case if the type of the key is inferred as an array, you may not have a reliable signature of the query function. It will be essentially something like this `(...keysOrVariablesOrMore: (TKey[number] | TVariables[number] | TMoreVariable | undefined)[]) => Promise<TResult>`.
So I chose to have a reasonable limitation to always use tuples up to some size. We can increase the maximum size at any time.

I actually experimented just now how to enforce TS to infer tuples instead of arrays. It seems with a few tricks it is possible to have any tuple size. I'll update the code shortly.

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: So, unfortunately, we cannot do this for `error` either. This is because the type of error is not `Error` (built-in class) but any possible throwable type (the library does not wrap inner exceptions), which can be anything, including primitives, `null` or `undefined`. That makes the type be `unknown | null` which is just `unknown`. So any CFA based on that does not make sense.

Comment (Igorbek) [reply types/react-query/index.d.ts]: To clarify, `const { status, error } = useQuery(key, () => throw null)` would produce `status === 'error' && error === null`.

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: I think we should follow the [API docs](https://github.com/tannerlinsley/react-query#usequery) and not the internal implementation.

**Comment by Igorbek** [comment]
@lukaszfiszer thank you for the feedback, and sorry again that I jumped on your work.

I'm addressing the concerns. It seems I can do something with the discriminated unions, let me see what I can do.

Regarding the `typescript-tuple`, I chose to use `ts-toolbelt` because it is in the list of approved external packages of DT, but not `typescript-tuple`. The TS requirement is coming from this typing itself because it uses the recursive type reference. I can lower TS requirement if I change the type of `QueryKeyPart`:

```diff
- export type QueryKeyPart = string | object | boolean | number | null | readonly QueryKeyPart[] | null | undefined;
+ export type QueryKeyPart = string | object | boolean | number | null | readonly unknown[] | null | undefined;
```

This will make the key type is a bit less strict.

File: types/react-query/index.d.ts
Comment (lukaszfiszer) [reply types/react-query/index.d.ts]: But in the every definition of query function you're using `List.Concat` to create a type of args that will be passed to function:

``` type QueryFunctionWithVariables<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables> = (
    ...key: _.List.Concat<TKey, TVariables>
) => Promise<TResult>;
```

Unless I'm not seeing something here, this will also work if `TKey` and `TVars` are constrained to open-ended tuples (`[string, ...any[]]`)

**Comment by lukaszfiszer** [comment]
> @lukaszfiszer thank you for the feedback, and sorry again that I jumped on your work.

No worries, I'm glad that you took it over. 

> Regarding the `typescript-tuple`, I chose to use `ts-toolbelt` because it is in the list of approved external packages of DT, but not `typescript-tuple`. The TS requirement is coming from this typing itself because it uses the recursive type reference. I can lower TS requirement if I change the type of `QueryKeyPart`:

I wasn't aware that `ts-toolbelt` is on a recommended list. Then I'm fine with leaving it as it is

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: Added discriminated union that allows to have `status === 'success' && data !== TResult`

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: Added. [Source code](https://github.com/tannerlinsley/react-query/blob/v1.1.4/src/useMutation.js#L29-L49) is even weirder there, it just not have `error`, `data` in some states. The definition now follows actual behaviour.

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: If the key is open-ended, then you can say in which parameter position variables or fetch more variable will land. The latest code has commented out example where it fails. `_.List.Concat` cannot concat tuples if the left is open-ended: `_.List.Concat<[1, ...2[]], [3, 4]> = [3, 4]` when actually should be `[1, 2 | 3, 2| 4, ...2[]]`.
I found a way to do it with `typescript-tuple` but it needs to be approved first (https://github.com/microsoft/types-publisher/pull/758).

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: done

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts]: ok, fair enough. I changed.

**Comment by Igorbek** [comment]
Please have another look, as I addressed all the feedback.
I think it is now good to merge, let's wait for approvals.
There's a space for improvements, but it can be done in other PRs, like adding documentation and reorganizing a litle bit.

File: types/react-query/index.d.ts
Comment (dbertella) [inline types/react-query/index.d.ts]: Isn't it all right to have type Error here? Can it be any other type than that?

Comment (bjarkehs) [reply types/react-query/index.d.ts]: I think @Igorbek explained it quite well here: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43438/files#r399837161

> To clarify, `const { status, error } = useQuery(key, () => throw null)` would produce `status === 'error' && error === null`.

Comment (dbertella) [reply types/react-query/index.d.ts]: yeah but come on, this is very bad, later on in the code is quite useles, to do something like error.message that is what we all probably going to do you need to cast it to Error 
```
<pre>{(error as Error)?.message}</pre>
```
Something like this 😭

Comment (dbertella) [reply types/react-query/index.d.ts]: actually maybe `Error | null` can work fine?

Comment (Igorbek) [reply types/react-query/index.d.ts]: I understand the concern and wish thrown exceptions were always of type `Error`, but in JavaScript you can throw and catch anything, including primitives, `null`, `undefined`. This the why when you write `try { ... } catch (err) { ... }` TS doesn't give you the type of the error and just uses `any`. Most of the libraries that propagate exceptions have to decide what to use for error `any` or `unknown`. For example, promises always have `error: any` as a parameter to `onReject`. I chose to use `unknown` here as it was more conservative and puts further decisions on consumers.
`Error | null` wouldn't work either for the same reasons.

Comment (dbertella) [reply types/react-query/index.d.ts]: got it! Sad, but thanks for the explenation

Comment (bjarkehs) [reply types/react-query/index.d.ts]: I see where you are coming from though. In many other languages async functionality is represented with a result type, for instance: Result<SuccessType, ErrorType = unknown>, that way you can type both success and failure state. It could even default to unknown for the failure state.

Comment (bjarkehs) [reply types/react-query/index.d.ts]: However, unless this is blocking valid use cases (which I doubt since it simply requires you to cast to a type) then I think this should be added in a patch or minor version later on. I rather have non blocking functional types now that we can extend than 100% correct types in a month.

**Comment by Igorbek** [comment]
@lukaszfiszer @bjarkehs @dbertella @burtyish if all good for now, you can approve the PR and I can go ahead and merge it. Any other improvements will be done later and any help is very welcomed.

**Comment by tatchi** [comment]
Turns out that `data` can apparently be `undefined` even when the `status` is `success`.

It happens when we pass a falsy value as first argument. 

Repro: https://codesandbox.io/s/react-query-success-status-8jo51

You'll need to export a zip I guess because typescript definitions within the `d.ts` files don't load apparently on codesandbox

**Comment by tatchi** [comment]
Also if I'm not mistaken, `data` cannot be `undefined` when using the `suspense` config. Since the `loading` state would be handled by the nearest `React.Suspense` component.

**Comment by Igorbek** [comment]
Hm, it's interesting about `data`. I think these caveats would make types harder to use, and must be handled by a consumer. If we always set `data` to be `TResult | undefined` it'd be harder to use with control flow analysis... on the other hand, it should enforce people to check for `undefined` which might be better.
I don't know what'd be better. I would keep it as is for now, but if the majority of reviewers would think opposite, we'd change.

File: types/react-query/index.d.ts
Comment (bjarkehs) [inline types/react-query/index.d.ts]: Wouldn't it be enough to simply expand this and line 26 to be `TResult | undefined` (maybe `null` too?) ? At least 33 would not necessarily result in undefined, since it's not a falsy value. Then maybe this requires a bit more from the user but at least it shouldn't leave the user in a scenario where things break.

Comment (bjarkehs) [reply types/react-query/index.d.ts]: Oh nevermind, this is not enough. :disappointed:

**Comment by lukaszfiszer** [comment]
@Igorbek approved! But before merging, please remove any redundant commented-out lines in tests like: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e5878591de48eac41b7e0f444e18905dd3740ed4/types/react-query/react-query-tests.ts#L172 https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e5878591de48eac41b7e0f444e18905dd3740ed4/types/react-query/react-query-tests.ts#L190  https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e5878591de48eac41b7e0f444e18905dd3740ed4/types/react-query/react-query-tests.ts#L199

and also cleanup the `pr42830()` function - these will be confusing for any future contributors.

**Review by lukaszfiszer — APPROVED** [review]

**Comment by Igorbek** [comment]
> @Igorbek approved! But before merging, please remove any redundant commented-out lines in tests like: ...
> and also cleanup the `pr42830()` function - these will be confusing for any future contributors.

All done.

File: types/react-query/index.d.ts
Comment (dbertella) [reply types/react-query/index.d.ts]: Result ♥️! Anyway no problem at all! Thanks for merging this!

File: types/react-query/index.d.ts
Comment (mustafawm) [inline types/react-query/index.d.ts:139]: Hi @Igorbek 
Thanks for all the work you've done.
Should we include/allow `string` queryKeys here? The library seems to [allow it](https://github.com/tannerlinsley/react-query#string-only-query-keys)

File: types/react-query/index.d.ts
Comment (mustafawm) [inline types/react-query/index.d.ts:136]: I meant to comment 👆

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts:139]: You're welcome.
It is allowed, but not in the query function (the one that fetches data), it is allowed in `useQuery` and friends. Query functions, on the other hand, always accept the keys as set of arguments, no matter is that single string or an array.

**Review by mwmcode — COMMENTED** [review]
Apologies if this is not the place but that's where it's complaining for me

Screen Shot 2020-04-01 at 2 19 37 pm

I can switch it to `[url]` 

Thanks for your quick reply.

File: types/react-query/index.d.ts
Comment (Igorbek) [reply types/react-query/index.d.ts:136]: ah, ok, it might be a valid concern.
It has to be a different overload though. Because in this case, we have to infer the type of the key as `[T]` where `T extends string`. I will make an update soon.

Comment (Igorbek) [reply types/react-query/index.d.ts:136]: addressed here #43568

**Pull request merged**
